### PR TITLE
MGDOBR-989 Update Bridge/Processor from Shard only for READY status

### PR DESCRIPTION
**Pull request by sutaakar**
Signed-off-by: Karel Suta <ksuta@redhat.com>

[MGDOBR-989](https://issues.redhat.com/browse/MGDOBR-989)

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java [hunk 1]
```diff
@@ -87,11 +89,17 @@ public void createBridgeExecutor(ProcessorDTO processorDTO) {
             // create or update the secrets for the bridgeExecutor
             createOrUpdateBridgeExecutorSecret(bridgeExecutor, processorDTO);
         } else {
-            LOGGER.info("BridgeExecutor '{}' already exists. Notifying manager that it is ready.", processorDTO.getId());
-            processorDTO.setStatus(ManagedResourceStatus.READY);
-            managerClient.notifyProcessorStatusChange(processorDTO).subscribe().with(
-                    success -> LOGGER.debug("Ready notification for BridgeExecutor '{}' has been sent to the manager successfully", processorDTO.getId()),
-                    failure -> LOGGER.error("Failed to send updated status to Manager for entity of type '{}'", ProcessorDTO.class.getSimpleName(), failure));
+            Boolean isBridgeExecutorReady = existing.getStatus()
```
Comment (manstis) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java:92]: I believe `existing.getStatus().isReady()` does this already.

I was doing much the same here too... but your PR has beaten me to it; thank-you.

File: shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceImpl.java [hunk 2]
```diff
@@ -72,11 +74,17 @@ public void createBridgeIngress(BridgeDTO bridgeDTO) {
             // create or update the secrets for the bridgeIngress
             createOrUpdateBridgeIngressSecret(bridgeIngress, bridgeDTO);
         } else {
-            LOGGER.info("BridgeIngress '{}' already exists. Notifying manager that it is ready.", bridgeDTO.getId());
-            bridgeDTO.setStatus(ManagedResourceStatus.READY);
-            managerClient.notifyBridgeStatusChange(bridgeDTO).subscribe().with(
-                    success -> LOGGER.debug("Ready notification for BridgeIngress '{}' has been sent to the manager successfully", bridgeDTO.getId()),
-                    failure -> LOGGER.error("Failed to send updated status to Manager for entity of type '{}'", BridgeDTO.class.getSimpleName(), failure));
+            Boolean isBridgeIngressReady = existing.getStatus()
```
Comment (manstis) [inline shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceImpl.java:77]: Likewise (`existing.getStatus().isReady()`)

**Review by manstis — COMMENTED** [review]
There are other scenarios to possibly consider too.

This code-branch is ran if the `operator` dies before completing provisioning a CRD and notifying the `manager` of the final status. The final status could equally be `FAILED` so IMO we aught to cover that too.

This scenario will be more visible when [MGDOBR-52](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/968/files#diff-194ed6e590f594580e290c528efd6f511354890c0ef569fd70ca48366257bd7eR199-R207) is merged. If k8s fails to provision the CRD we're setting its `Status` to `READY:False`, `AUGMENTING:False`.

I'll close https://issues.redhat.com/browse/MGDOBR-988 as a duplicate.

**Comment by sutaakar** [comment]
good point, will cover failing scenario too

**Comment by manstis** [comment]
> good point, will cover failing scenario too

It is a can of worms though @sutaakar something we might need to get the almighty @r00ta comment on before his PTO.

When the reconciliation loop first starts the status is `READY:Unknown`/`AUGMENTING:Unknown` by virtue of the [constructor](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/BridgeExecutor.java#L184) used to build the [_expected_](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceImpl.java#L73) CRD specification.

Subsequently and depending on various other [conditions](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/CustomResourceStatus.java#L54-L73) we  set the status to different permutations as provisioning progresses.

We however use `READY:False`/`AUGMENTING:False` to represent both a **failed** state ([timeout](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/CustomResourceStatus.java#L63-L64) and [replica failure](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/resources/CustomResourceStatus.java#L66-L67)) and an [intermediate](subsequently) **provisioning** state. Once the _deployment_ is ready we set the state to `READY:False`/`AUGMENTING:True` in the next [phase](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/main/java/com/redhat/service/smartevents/shard/operator/controllers/BridgeExecutorController.java#L117-L118) of provisioning.

We should probably consider making the Status `READY:False`/`AUGMENTING:True` for when the deployment is not ready and keep `READY:False`/`AUGMENTING:False` to represent a failed state (or add a new `ConditionTypeConstant` altogether to represent failure....). This decision won't be needed really until MGDOBR-52 is closer to being merged (as that's when `READY:False`/`AUGMENTING:False` represents absolute failure!)

**Comment by sutaakar** [comment]
@manstis IMHO we don't need to address intermediate states, because these should be covered by reconciliation loops, though there is a question how to distinguish it from failed state.
I can cover just READY state in this PR, leaving the failing state on you as part of the 52?

**Comment by manstis** [comment]
> @manstis IMHO we don't need to address intermediate states, because these should be covered by reconciliation loops, though there is a question how to distinguish it from failed state. I can cover just READY state in this PR, leaving the failing state on you as part of the 52?

Yes @sutaakar :smiley: that's fine.. the extended scenario does not exist (yet).

**Comment by manstis** [comment]
The failing tests _may_ be transitive failures.. let's keep an eye on them.

When I was writing my equivalent change here I noticed transient (new) test failures because k8s continued with the reconciliation loop when we were trying to make assertions on properties k8s may have updated depending on who wins the race (to execute code first). 

If they keep failing an alternative test may be needed.

**Comment by manstis** [comment]
Looks like I need to reconsider the tests...

**Comment by sutaakar** [comment]
Doing some experiments with tests....

**Comment by manstis** [comment]
See:
- https://issues.redhat.com/browse/MGDOBR-991
- https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1025

**Comment by sutaakar** [comment]
@manstis Adjusted implementation.
Was thinking about test coverage for the change......
One option is to immediately call `bridgeExecutorService.createBridgeExecutor(dto);` second time right after first one when we could expect that k8s reconciler didn't create everything needed. But still it would be a sort of race condition as we would test intermediate state. Or the additional coverage would be provided with implementation of failure conditions handling later.

**Review by manstis — APPROVED** [review]

**Comment by Srihari1192** [comment]
@sutaakar Noticed below error once while Provisioning/Deleting bridge in Fleet Shard

`WARN  [com.red.ser.sma.sha.ope.ManagerSyncServiceImpl] (vert.x-eventloop-thread-3) Provisioning of Bridge 'b7a54a95-b6f7-4fe2-a87a-97b088851880' failed: java.lang.NullPointerException: [BridgeIngress] Name can't be null
        at java.base/java.util.Objects.requireNonNull(Objects.java:246)
        at com.redhat.service.smartevents.shard.operator.resources.BridgeIngress$Builder.validate(BridgeIngress.java:145)
        at com.redhat.service.smartevents.shard.operator.resources.BridgeIngress$Builder.build(BridgeIngress.java:117)
        at com.redhat.service.smartevents.shard.operator.resources.BridgeIngress.fromDTO(BridgeIngress.java:63)
        at com.redhat.service.smartevents.shard.operator.BridgeIngressServiceImpl.createBridgeIngress(BridgeIngressServiceImpl.java:59)
        at com.redhat.service.smartevents.shard.operator.BridgeIngressServiceImpl_Subclass.createBridgeIngress$$superforward1(Unknown Source)
        at com.redhat.service.smartevents.shard.operator.BridgeIngressServiceImpl_Subclass$$function$$5.apply(Unknown Source)
`

`WARN  [com.red.ser.sma.sha.ope.ManagerSyncServiceImpl] (vert.x-eventloop-thread-3) Deleting of Bridge 'b7a54a95-b6f7-4fe2-a87a-97b088851880' failed: java.lang.NullPointerException: [BridgeIngress] Name can't be null
        at java.base/java.util.Objects.requireNonNull(Objects.java:246)
        at com.redhat.service.smartevents.shard.operator.resources.BridgeIngress$Builder.validate(BridgeIngress.java:145)
        at com.redhat.service.smartevents.shard.operator.resources.BridgeIngress$Builder.build(BridgeIngress.java:117)
        at com.redhat.service.smartevents.shard.operator.resources.BridgeIngress.fromDTO(BridgeIngress.java:63)
        at com.redhat.service.smartevents.shard.operator.BridgeIngressServiceImpl.deleteBridgeIngress(BridgeIngressServiceImpl.java:91)
        at com.redhat.service.smartevents.shard.operator.BridgeIngressServiceImpl_Subclass.deleteBridgeIngress$$superforward1(Unknown Source)`

**Comment by sutaakar** [comment]
@Srihari1192 sorry, I didn't reproduce your issue.
Can you check whether you built the whole repo first before starting manager and shard?

**Comment by Srihari1192** [comment]
> @Srihari1192 sorry, I didn't reproduce your issue. Can you check whether you built the whole repo first before starting manager and shard? Do you have empty database when starting the manager?

I didn't able to reproduce issue again , it happened once when bridge failed. 
Let me try by emptying database

**Comment by manstis** [comment]
@sutaakar 

> Was thinking about test coverage for the change......

The revised [tests](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceTest.java#L163-L183) ensure ALL dependencies have been provisioned by k8s before our reconciler runs. This ensures the `BridgeIngress` is `READY` before we make the [subsequent](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceTest.java#L197) call and assert [ManagerClient](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceTest.java#L199) was invoked.

To test your change you could duplicate the existing test but remove one of the pre-provisioning steps. This would mean the `BridgeIngress` was not transitioned into `READY` and you could assert `ManagerClient` was not invoked.

**Comment by sutaakar** [comment]
@manstis test coverage added

**Review by manstis — APPROVED** [review]
LGTM :+1: Thank-you for the tests.

**Comment by Srihari1192** [comment]
> > @Srihari1192 sorry, I didn't reproduce your issue. Can you check whether you built the whole repo first before starting manager and shard? Do you have empty database when starting the manager?
> 
> I didn't able to reproduce issue again , it happened once when bridge failed. Let me try by emptying database

Seems random issue due to environment..Didn't repropduced again

**Review by Srihari1192 — APPROVED** [review]

**Review by r00ta — APPROVED** [review]

**Pull request merged**